ktor-server/ktor-server-plugins/ktor-server-websockets/jvm/test/io/ktor/tests/websocket/WebSocketTest.kt

```diff
@@ -118,593 +106,402 @@ class WebSocketTest {
             webSocket("/echo") {
                 val data = receiveDeserialized<Data>(typeInfo<Data>())
                 sendSerialized(data, typeInfo<Data>())
-                outgoing.send(Frame.Close())
             }
         }
 
         val jsonData = "[hello]"
 
-        createClient {
-            install(io.ktor.client.plugins.websocket.WebSockets)
-        }.webSocket("/echo") {
-            outgoing.send(Frame.Text(jsonData))
-
-            val frame = incoming.receive() as Frame.Text
-            assertEquals(FrameType.TEXT, frame.frameType)
-            assertEquals(jsonData, frame.readText())
+        createWebSocketsClient().webSocket("/echo") {
+            send(jsonData)
+            assertEquals(jsonData, receiveText())
         }
     }
 
     @Test
-    fun testSerializationWithNoConverter() {
-        withTestApplication {
-            application.install(WebSockets) {}
-
-            application.routing {
-                webSocket("/echo") {
-                    assertFailsWith<WebsocketConverterNotFoundException>("No converter was found for websocket") {
-                        receiveDeserialized<Data>()
-                    }
-                    assertFailsWith<WebsocketConverterNotFoundException>("No converter was found for websocket") {
-                        sendSerialized(Data("hello"))
-                    }
-                    outgoing.send(Frame.Close())
-                }
-            }
-
-            val jsonData = "[hello]"
-            val sendBuffer = ByteBuffer.allocate(jsonData.length + 100)
-
-            Serializer().apply {
-                enqueue(Frame.Text(jsonData))
-                serialize(sendBuffer)
-            }
+    fun testSerializationWithNoConverter() = testApplication {
+        install(WebSockets)
 
-            val conversation = Job()
-
-            handleWebSocket("/echo") {
-                bodyChannel = writer {
-                    channel.writeFully(sendBuffer.array())
-                    channel.flush()
-                    conversation.join()
-                }.channel
-            }.let {
-                runBlocking {
-                    withTimeout(Duration.ofSeconds(10).toMillis()) {}
+        val executed = Job()
+        routing {
+            webSocket("/echo") {
+                assertFailsWith<WebsocketConverterNotFoundException>("No converter was found for websocket") {
+                    receiveDeserialized<Data>()
+                }
+                assertFailsWith<WebsocketConverterNotFoundException>("No converter was found for websocket") {
+                    sendSerialized(Data("hello"))
                 }
+                executed.complete()
```

Comment:
> "What happens on a server stays on the server" (c)

Assertion failures on a server are caught and logged by the server, so they don't affect the test result. Is there a recommended way to make assertions on a server? My current solution produces a timeout if one of the assertions fails, because `executed.complete()` will never be called, but I don't like that we miss actual assertion errors.

Reply:
It'd be better to catch and send the errors from the server I would think.  Normally, you could add exception handling with https://ktor.io/docs/server-status-pages.html but I guess for websockets you'd want to send a custom message or something.

Reply:
Status pages plugin is not added as a dependency here, so I replaced `Job` with `Deferred<Result<Unit>>`. Could you [take a look](https://github.com/ktorio/ktor/pull/4295/commits/6af7ca5bce5c12060462c89317c569ccb768dd12)?